packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -1943,6 +1944,51 @@ export class MergeTree {
         }
     }
 
+    /**
+     * Perform local rollback of previously deleted segments
+     */
+    public rollbackRemovedSegment(segments: ISegment[]) {
+        for (const segment of segments) {
+            segment.removedClientIds = undefined;
+            segment.removedSeq = undefined;
+            segment.localRemovedSeq = undefined;
+
+            if (this.mergeTreeDeltaCallback) {
+                const insertOp = createInsertSegmentOp(this.findRollbackPosition(segment), segment);
+                this.mergeTreeDeltaCallback(
+                    { op: insertOp },
+                    {
+                        operation: MergeTreeDeltaType.INSERT,
+                        deltaSegments: [{ segment }],
+                    });
+            }
+        }
+
+        this.blockUpdate(this.root);
```

Comment:
`blockUpdate` isn't recursive, and it's probably not ok to make it so here. Other logic generally calls this on ascending nodes of the tree in the modifying walks. I can help you think through what a test for this scenario should look like later today if this isn't enough to go on.

For consistency, I think it's probably better to call `this.blockUpdateLengths` here with `UnassignedSeq` and the local client id. It will end up doing the same thing, but most other tree-affecting walks use that method (which correctly handles partial lengths when it needs to).

I need to do some more thinking on if this interacts poorly with any of the planned changes I had for partial lengths.... I think it's probably ok, but the area is a bit delicate.

Reply:
I'm imagining something like this instead for the implementation of this method:
```
for (segment of segments) {
    // same undo remove logic
    // same invoke tree callback

    for (let updateNode = segment.parent; updateNode !== undefined; updateNode = updateNode.parent) {
        this.blockUpdateLengths(updateNode, UnassignedSequenceNumber, localClientId);
    }
}
```

Reply:
Do I need to add anything further to test this? I don't fully understand this code.

(I didn't find any existing helper for walking up the tree.)

Reply:
Might be worth adding a test that the old version failed on. The impact of not walking up is that levels of the tree further from the leaf will have incorrect leaves--they'll "think" they have length equal to post-remove, but they should actually have length pre-remove (or post-rollback). This is only evident if you perform some operation that depends on that length being correct, though. Probably the clearest way to check is to construct a tree which has enough length 1 segments to have 3 layers of nodes, then remove + rollback a segment in the first 8 or so characters, then try and insert something near the end. The inserting walk should skip over `root.children[0]`'s children after looking at `root.children[0]`'s cached length, so if that cached length is incorrectly computed you'll end up with your text in the wrong place.

Reply:
I verified it fails with the earlier version and passes now.